### PR TITLE
add-submission-mode-to-db

### PR DESCRIPTION
**Pull request by oanaBiancaSchiopu**

### CODE REVIEW

File: rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/data/DataUtility.java [hunk 1]
```diff
@@ -283,6 +285,12 @@ public static FilteredStatistics toFilteredStatistics(FilteredStatisticsData job
                                       jobsReportData.getSuccessfulRate());
     }
 
+    public static Map<String, String> setSubmissionModeToGenericInfo(Map<String, String> genericInformation) {
```
Comment (fviale) [inline rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/data/DataUtility.java:288]: careful, provided generic info can be null

File: rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/SchedulerClient.java [hunk 2]
```diff
@@ -933,6 +933,7 @@ public JobId submit(Job job)
             throws NotConnectedException, PermissionException, SubmissionClosedException, JobCreationException {
         JobIdData jobIdData = null;
         try {
+            job.setGenericInformation(DataUtility.setSubmissionModeToGenericInfo(job.getGlobalGenericInformation()));
```
Comment (fviale) [inline rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/SchedulerClient.java:936]: why externalize this method in DataUtility as it's apparently used only inside SchedulerClient (unless I missed one use) ?

File: rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/data/DataUtility.java [hunk 3]
```diff
@@ -283,6 +285,12 @@ public static FilteredStatistics toFilteredStatistics(FilteredStatisticsData job
...
+        Map<String, String> newGenericInformation = new LinkedHashMap<>(genericInformation);
```
Comment (fviale) [inline rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/data/DataUtility.java:289]: "scheduler-api" should be externalized as a constant, typically in SchedulerConstants

File: rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/data/DataUtility.java [hunk 4]
```diff
@@ -29,7 +29,9 @@
 
 import java.io.IOException;
 import java.util.ArrayList;
+import java.util.LinkedHashMap;
 import java.util.List;
+import java.util.Map;
```
Comment (fviale) [inline rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/data/DataUtility.java:34]: in this class, the toJobInfo(..) method should also be modified to include the submissionMode

File: rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/SchedulerClient.java [hunk 5]
```diff
@@ -1842,6 +1851,13 @@ private Optional<String> getRevisionFromCalledWorkflow(String calledWorkflow) {
         }
     }
 
+    public static Map<String, String> setSubmissionModeToGenericInfo(Map<String, String> genericInformation) {
+        Map<String, String> newGenericInformation = genericInformation == null ? new LinkedHashMap<>()
+                                                                               : new LinkedHashMap<>(genericInformation);
+        newGenericInformation.put("submission.mode", SCHEDULER_API);
```
Comment (fviale) [inline rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/SchedulerClient.java:1857]: sorry, I forgot to mention "submission.mode" should also be defined as a global constant

File: scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/SchedulerConstants.java [hunk 6]
```diff
@@ -183,4 +183,9 @@ public class SchedulerConstants {
      */
     public static String TASK_WALLTIME_GENERIC_INFO = "WALLTIME";
 
+    /**
+     * This generic information can be used to set the portal name via the job was submitted
+     */
+    public static String SCHEDULER_API = "scheduler-api";
```
Comment (fviale) [inline scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/SchedulerConstants.java:189]: SUBMISSION_MODE_SCHEDULER_API ?

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulerFrontendState.java [hunk 7]
```diff
@@ -541,6 +541,9 @@ InternalJob createJob(Job userJob, UserIdentificationImpl ident)
         job.setTenant(ident.getTenant());
         // route project name inside job info
         job.setProjectName(job.getProjectName());
+        if (job.getGenericInformation() != null) {
+            job.setSubmissionMode(job.getGenericInformation().get("submission.mode"));
```
Comment (fviale) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulerFrontendState.java:545]: please use a global constant

**Review by fviale — APPROVED** [review]
LGTM

**Comment by mboussaa** [comment]
I don't understand how it works; the submission mode is only known through generic infos?

**Comment by fviale** [comment]
@mboussaa yes, it is the only way

**Comment by mboussaa** [comment]
what are the possible values of submission mode? and do we have a default one?

**Review by AlexFalessi — APPROVED** [review]

**Comment by fviale** [comment]
@mboussaa 

no default value (can be null), and possible values are not fixed.

File: rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/SchedulerClient.java [hunk 8]
```diff
@@ -1842,6 +1852,13 @@ private Optional<String> getRevisionFromCalledWorkflow(String calledWorkflow) {
         }
     }
 
+    public static Map<String, String> setSubmissionModeToGenericInfo(Map<String, String> genericInformation) {
+        Map<String, String> newGenericInformation = genericInformation == null ? new LinkedHashMap<>()
+                                                                               : new LinkedHashMap<>(genericInformation);
+        newGenericInformation.put(SUBMISSION_MODE, SUBMISSION_MODE_SCHEDULER_API);
```
Comment (fviale) [inline rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/SchedulerClient.java:1858]: could you check if SUBMISSION_MODE is not already in the generic info, if yes, then don't set it

**Pull request merged**